splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/UpgradeFixIndexDescriptors.java

```diff
@@ -0,0 +1,85 @@
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.derby.impl.sql.catalog.upgrade;
+
+import com.splicemachine.access.api.PartitionAdmin;
+import com.splicemachine.access.configuration.HBaseConfiguration;
+import com.splicemachine.db.catalog.IndexDescriptor;
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.sql.dictionary.ConglomerateDescriptor;
+import com.splicemachine.db.iapi.sql.dictionary.DataDictionary;
+import com.splicemachine.db.iapi.sql.dictionary.SchemaDescriptor;
+import com.splicemachine.db.iapi.sql.dictionary.TableDescriptor;
+import com.splicemachine.db.iapi.store.access.TransactionController;
+import com.splicemachine.db.impl.sql.catalog.BaseDataDictionary;
+import com.splicemachine.derby.impl.sql.catalog.SpliceDataDictionary;
+import com.splicemachine.derby.impl.sql.catalog.Splice_DD_Version;
+import com.splicemachine.derby.impl.sql.execute.operations.ScanOperation;
+import com.splicemachine.derby.impl.store.access.SpliceTransactionManager;
+import com.splicemachine.derby.impl.store.access.base.SpliceConglomerate;
+import com.splicemachine.derby.utils.ConglomerateUtils;
+import com.splicemachine.si.api.txn.Txn;
+import com.splicemachine.si.impl.driver.SIDriver;
+import com.splicemachine.utils.SpliceLogUtils;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+public class UpgradeFixIndexDescriptors extends UpgradeScriptBase {
+    public UpgradeFixIndexDescriptors(SpliceDataDictionary sdd, TransactionController tc) {
+        super(sdd, tc);
+    }
+
+    @Override
+    protected void upgradeSystemTables() throws StandardException {
+        for (TableDescriptor td: sdd.getAllTableDescriptors()) {
+            SchemaDescriptor sd = td.getSchemaDescriptor();
+            for (ConglomerateDescriptor cd: td.getConglomerateDescriptors()) {
+                if (cd.isIndex()) {
+                    fixIndexDescriptorColumns(td, sd, cd);
+                }
+                addKeyFormatIdsToConglomerate(cd);
+            }
+        }
+    }
+
+    private void fixIndexDescriptorColumns(TableDescriptor td, SchemaDescriptor sd, ConglomerateDescriptor cd) throws StandardException {
+        IndexDescriptor id = cd.getIndexDescriptor();
+        if (id.isInvalidIndexDescriptor()) {
```

Comment:
According to my previous index on expressions experience, `id` can be `null` for some conglomerate descriptors. Also, even `id` is not null, its internal `id` might be null. Maybe we should guard that?

Here is an example code snippet that testing something similar in `PredicateList.java`:
```
IndexRowGenerator irg = cd == null ? null : cd.getIndexDescriptor();
if (irg != null && irg.getIndexDescriptor() == null) {
    irg = null;
}
```

Reply:
I think IndexDescriptor cannot be null here, because we check for `cd.isIndex()` before.